### ISSUE TITLE
Feature Request: decompress from input `Iterator` to output `Iterator` (for low RAM usage in embedded Rust)

### ISSUE THREAD

**Issue by neil-morrison44** [Frommi/miniz_oxide#140]
Hi,

Feel free to point me in the right direction if this is supported just now (or is implemented in another crate I've not found yet), but what I'm looking for is a function with an input similar to `decompress_slice_iter_to_slice` but that returns an iterator I can iterate over to get decompressed data. I'm trying to run on RAM constrained devices in `no_std`, eventually reading image files and writing pixels directly to the display.

Ideally, and I might be far off here from how it'd actually need to work:
- the `decompress_*` function gets called & is given an `Iterator` that returns `u8`s
- it returns a struct implementing `core::iter::Iterator` over `u8`s
- calling `next()` on the returned iter will fill a (small) internal buffer by calling the input `Iterator`, decompressing, then transparently returning values from the output buffer until it's consumed, refilling it as needed

Ideally the size of the internal buffer would be definable by the user (const generic?, `heapless` does something similar) within whatever bounds make sense.

This is potentially related to https://github.com/Frommi/miniz_oxide/issues/25 but more targeted towards resource-constrained devices (embedded Rust etc)

**Comment by oyvindln** [comment]
Yeah it seems reasonable to have something like this for cases where flate2 is too heavy. I've not personally done much work on this library in quite a while, might look into it at some point if no one else starts tinkering with it but might take a while.

**Comment by neil-morrison44** [comment]
I'm still thinking about this one occasionally - I _think_ it'd be fairly easy/nice to do with a generator which'll be available in the 2024 edition. Since iterator in / iterator out things should become really neat with generators.

I'll probably have free some time at the end of the year to look into it, unless anyone else wants to. 
(Which'll align with the 2024 edition's planned release in early 2025)

**Comment by neil-morrison44** [comment]
For embedded things knowing that you'll require a 32k buffer ahead of time is completely fine (unless you're building for a chip with less than 32k obviously) since that's a size known at compile time.

It's the input buffer & output buffer being unknown sizes that makes embedded use impossible, assuming a use case like reading a compressed image from an SD card and drawing to a display or reading compressed sound from a network socket & playing it on a speaker.

**Comment by peckpeck** [comment]
In this case it is easy to develop, here is how I would do it https://github.com/Frommi/miniz_oxide/commit/29d3a0f4ca21064825bb6c25c308856e5081761f

Please note 2 things:
* there is no error handling inside iterators, so an error is translated into the end of the stream
* the buffer is fixed and always 32k (this could be refactored)
